runtime/src/bank/builtins/core_bpf_migration/mod.rs

```diff
@@ -3,7 +3,391 @@ pub(crate) mod error;
 mod source_upgradeable_bpf;
 mod target_builtin;
 
+use {
+    crate::bank::Bank,
+    error::CoreBpfMigrationError,
+    solana_sdk::{
+        account::{AccountSharedData, ReadableAccount},
+        bpf_loader_upgradeable::UpgradeableLoaderState,
+        pubkey::Pubkey,
+    },
+    source_upgradeable_bpf::SourceUpgradeableBpf,
+    std::sync::atomic::Ordering::Relaxed,
+    target_builtin::TargetBuiltin,
+};
+
+/// Sets up a Core BPF migration for a built-in program.
+#[derive(Debug)]
 pub(crate) enum CoreBpfMigrationTargetType {
+    /// Builtin programs should have a program account.
     Builtin,
+    /// Stateless builtins should not have a program account.
     Stateless,
 }
+
+/// Configurations for migrating a built-in program to Core BPF.
+#[derive(Debug)]
+pub(crate) struct CoreBpfMigrationConfig {
+    /// The source program ID to replace the builtin with.
+    pub source_program_id: Pubkey,
+    /// The feature gate to trigger the migration to Core BPF.
+    /// Note: This feature gate should never be the same as any builtin's
+    /// `enable_feature_id`. It should always be a feature gate that will be
+    /// activated after the builtin is already enabled.
+    pub feature_id: Pubkey,
+    /// The type of migration to perform.
+    pub migration_target: CoreBpfMigrationTargetType,
+    pub datapoint_name: &'static str,
+}
+
+fn checked_add(a: usize, b: usize) -> Result<usize, CoreBpfMigrationError> {
+    a.checked_add(b)
+        .ok_or(CoreBpfMigrationError::ArithmeticOverflow)
+}
+
+/// Create a new `Account` with a pointer to the target's new data account.
+///
+/// Note the pointer is created manually, as well as the owner and
+/// executable values. The rest is inherited from the source program
+/// account, including the lamports.
+fn create_new_target_program_account(
+    target: &TargetBuiltin,
+    source: &SourceUpgradeableBpf,
+) -> Result<AccountSharedData, CoreBpfMigrationError> {
+    let state = UpgradeableLoaderState::Program {
+        programdata_address: target.program_data_address,
+    };
+    let data = bincode::serialize(&state)?;
+    // The source program account has the same state, so it should already have
+    // a sufficient lamports balance to cover rent for this state.
+    // Out of an abundance of caution, first ensure the source program
+    // account's data is the same length as the serialized state.
+    if source.program_account.data().len() != data.len() {
+        return Err(CoreBpfMigrationError::InvalidProgramAccount(
+            source.program_address,
+        ));
+    }
+    // Then copy the source account's contents and overwrite the data with the
+    // newly created target program account data.
+    let mut account = source.program_account.clone();
+    account.set_data_from_slice(&data);
+    Ok(account)
+}
+
+impl CoreBpfMigrationConfig {
+    pub(crate) fn migrate_builtin_to_core_bpf(
```

Comment:
An argument could be made to move it into Bank.

However, I would counter by saying:

- Bank is large enough as it is.
- Any helpers, enums, etc. that need to be used by this method can live in this submodule, privately.
- The checked account loader modules (`mod target_builtin`) can remain private.
- The method itself is only available to Bank when a `BuiltinPrototype` has `Some(config)` for its `core_bpf_migration_config` field (helps define explicit use).

I even went to some lengths in the original PR to apply a wrapper method to both `BuiltinPrototype` and `StatelessBuiltinPrototype`, so it was very difficult to fumble the target's program ID from within Bank, but Jon had suggested that was too much.

What do you think about leaving it where it is?

Reply:
To be clear, I'm not saying move this into the `bank.rs` file. I'm suggesting leave this where it is, but written as `impl Bank`.

The `core_bpf_migration` module is already imported into the `bank` module via `builtins`, and anything marked as `pub(crate)` is already visible in `bank`.

> - Any helpers, enums, etc. that need to be used by this method can live in this submodule, privately.
> - The checked account loader modules (mod target_builtin) can remain private.

These things remain true with the approach I am proposing.

> The method itself is only available to Bank when a BuiltinPrototype has Some(config) for its core_bpf_migration_config field (helps define explicit use).

Hmm, this method is available to any `CoreBpfMigrationConfig`, with a Bank to pass in. The fact that a given `BuiltinPrototype` does or does not specify a config doesn't actually limit a caller's ability to use the method.

Reply:
What I'm proposing is really a pretty small change. This is the extent of the (non-test) diff:
```diff
--- a/runtime/src/bank/builtins/core_bpf_migration/mod.rs
+++ b/runtime/src/bank/builtins/core_bpf_migration/mod.rs
@@ -82,16 +82,16 @@ fn new_target_program_account(
     Ok(account)
 }

-impl CoreBpfMigrationConfig {
+impl Bank {
     pub(crate) fn migrate_builtin_to_core_bpf(
-        &self,
-        bank: &mut Bank,
+        &mut self,
+        config: CoreBpfMigrationConfig,
         program_id: &Pubkey,
     ) -> Result<(), CoreBpfMigrationError> {
-        datapoint_info!(self.datapoint_name, ("slot", bank.slot, i64));
+        datapoint_info!(config.datapoint_name, ("slot", self.slot, i64));

-        let target = TargetBuiltin::new_checked(bank, program_id, &self.migration_target)?;
-        let source = SourceUpgradeableBpf::new_checked(bank, &self.source_program_id)?;
+        let target = TargetBuiltin::new_checked(self, program_id, &config.migration_target)?;
+        let source = SourceUpgradeableBpf::new_checked(self, &config.source_program_id)?;

         // Attempt serialization first before touching the bank.
         let new_target_program_account = new_target_program_account(&target, &source)?;
@@ -115,33 +115,33 @@ impl CoreBpfMigrationConfig {

         // Burn lamports from the target program account, since it will be
         // replaced.
-        bank.capitalization
+        self.capitalization
             .fetch_sub(target.program_account.lamports(), Relaxed);

         // Replace the target builtin account with the
         // `new_target_program_account` and clear the source program account.
-        bank.store_account(&target.program_address, &new_target_program_account);
-        bank.store_account(&source.program_address, &AccountSharedData::default());
+        self.store_account(&target.program_address, &new_target_program_account);
+        self.store_account(&source.program_address, &AccountSharedData::default());

         // Copy the source program data account into the account at the target
         // builtin program's data address, which was verified to be empty by
         // `TargetBuiltin::new_checked`, then clear the source program data
         // account.
-        bank.store_account(&target.program_data_address, &source.program_data_account);
-        bank.store_account(&source.program_data_address, &AccountSharedData::default());
+        self.store_account(&target.program_data_address, &source.program_data_account);
+        self.store_account(&source.program_data_address, &AccountSharedData::default());

         // Remove the built-in program from the bank's list of built-ins.
-        bank.builtin_program_ids.remove(&target.program_address);
+        self.builtin_program_ids.remove(&target.program_address);

         // Unload the programs from the bank's cache.
-        bank.transaction_processor
+        self.transaction_processor
             .program_cache
             .write()
             .unwrap()
             .remove_programs([source.program_address, target.program_address].into_iter());

         // Update the account data size delta.
-        bank.calculate_and_update_accounts_data_size_delta_off_chain(old_data_size, new_data_size);
+        self.calculate_and_update_accounts_data_size_delta_off_chain(old_data_size, new_data_size);

         Ok(())
     }
 ```

Reply:
Ahhh I see. I was imagining you meant move it into `bank.rs` under the rest of the huge `impl` there. This is cool with me, then. Agreed those first three points are preserved.

> Hmm, this method is available to any `CoreBpfMigrationConfig`, with a Bank to pass in. The fact that a given BuiltinPrototype does or does not specify a config doesn't actually limit a caller's ability to use the method.

Fair point. I figured unless someone deliberately instantiated a `CoreBpfMigrationConfig`, they would never encounter this method unless someone changed `BUILTINS` or provided the config on an additional builtin. With your suggestion, we're still in the same boat, since it's needed as a parameter.

Change made.